test/ebpf/testSockProg.c

```diff
@@ -0,0 +1,428 @@
+
+#include <stdlib.h>
+#include <string.h>
+#include "bpf/bpf.h"
+#include "bpf/libbpf.h"
+#include "testSockProg.h"
+
+char *get_map_pin_path(const char *map_name)
+{
+    char *map_pin_path = malloc(sizeof(char) * (strlen(map_name) + strlen(DEFAULT_MAP_PIN_PATH_PREFIX) + 1));
+    strcpy(map_pin_path, DEFAULT_MAP_PIN_PATH_PREFIX);
+    strcat(map_pin_path, map_name);
+    return map_pin_path;
+}
+
+void get_epmap_name(int internal_map_type, int comp_id, char **full_map_name)
+{
+    printf("in get_epmap_name func\n");
+    switch (internal_map_type)
+    {
+    case COMP_POLICY_MAP:
+    {
+        int prefixlen =  strlen(COMP_PMAP_NAME_PREFIX)+5;
+        char *numVal;
+        numVal = (char*) malloc(prefixlen*sizeof(char));
+        sprintf(numVal, "%s%d",COMP_PMAP_NAME_PREFIX, comp_id);        
+        *full_map_name = numVal;
+        break;
+    }
+    case GLOBAL_POLICY_MAP:
+        *full_map_name = GLOBAL_PMAP_NAME;
+        break;
+    case IP_CACHE_MAP:
+        *full_map_name = IP_CACHE_MAP_NAME;
+        break;
+    }
+}
+
+int pin_given_map(int internal_map_type, fd_t fd)
+{
+    char *map_name = NULL;
+    get_epmap_name(internal_map_type, POLICY_MAP_ID, &map_name);
+
+    printf("pin_given_map: map name: %s\n", map_name);
+    // Map fd is invalid. Open fd to the map.
+    char *pin_path = get_map_pin_path(map_name);
+    printf("pin_given_map: map pinned path %s\n", pin_path);
+    fd_t fdnew = bpf_obj_get(pin_path);
+    if (fdnew != INVALID_MAP_FD)
+    {
+        printf("pin_given_map: found the pinned map FD\n");
+        return 0;
+    }
+
+    printf("pin_given_map: pinned map not found, creating pin for map %s\n", map_name);
+    // Map created. Now pin the map.
+    int error = bpf_obj_pin(fd, pin_path);
+    if (error != 0)
+    {
+        // close map fd.
+        _close(fd);
```

Comment:
This looks like a bug.  One should not call _close on FD's returned from `bpf_map__fd()`.  Instead one should call `bpf_object__close()` on the object returned by `bpf_prog_load()` when done.   `bpf_map__fd()` just returns an FD already stored in the object.  The docs on https://microsoft.github.io/ebpf-for-windows/libbpf_8h.html#a37b9672527f744e7492797529f4f9a6a make this clear for the fd returned by bpf_prog_load but we can update the docs to be clear about the fds returned by bpf_map_fd too.

Reply:
FYI, documentation has now been updated

Reply:
Okay, will remove this call.